trunk/SUAVE/Methods/Aerodynamics/Airfoil_Panel_Method/airfoil_analysis.py

```diff
@@ -43,177 +69,303 @@ def airfoil_analysis(airfoil_geometry,alpha,Re_L,npanel = 100):
-    delta_star_top_func   = interp1d(x_ts[::-1] ,del_star_top[::-1])
-    delta_star_bot_func   = interp1d(x_bs,del_star_bot) 
-    H_top_func            = interp1d(x_ts[::-1],H_top[::-1])
-    H_bot_func            = interp1d(x_bs,H_bot)   
-    Cf_top_func           = interp1d(x_ts[::-1],cf_top[::-1])
-    Cf_bot_func           = interp1d(x_bs,cf_bot) 
-    delta_top_func        = interp1d(x_ts[::-1],delta_top[::-1])
-    delta_bot_func        = interp1d(x_bs,delta_bot)     
-                          
-    theta_vals            = np.concatenate([theta_top_func(x_top[::-1]),theta_bot_func(x_bot)])
-    delta_star_vals       = np.concatenate([delta_star_top_func(x_top[::-1]),delta_star_bot_func(x_bot)])
-    H_vals                = np.concatenate([H_top_func(x_top[::-1]),H_bot_func(x_bot)])
-    Cf_vals               = np.concatenate([Cf_top_func(x_top[::-1]),Cf_bot_func(x_bot)])
-    delta_vals            = np.concatenate([delta_top_func(x_top[::-1]),delta_bot_func(x_bot)])
+    airfoil_Cl       = np.zeros((nalpha,nRe))
+    airfoil_Cd       = np.zeros_like(airfoil_Cl)
+    airfoil_Cm       = np.zeros_like(airfoil_Cl)           
+    Cp_vals          = np.zeros((npanel,nalpha, nRe))
+    Xbl_vals         = np.zeros((npanel,nalpha, nRe))
+    Ybl_vals         = np.zeros((npanel,nalpha, nRe))
+    Ve_vals          = np.zeros_like(Cp_vals)
+    dVe_vals         = np.zeros_like(Cp_vals)
+    theta_vals       = np.zeros_like(Cp_vals)
+    delta_star_vals  = np.zeros_like(Cp_vals)
+    H_vals           = np.zeros_like(Cp_vals)
+    Cf_vals          = np.zeros_like(Cp_vals)
+    delta_vals       = np.zeros_like(Cp_vals)
+    Re_theta_t_vals  = np.zeros_like(Cp_vals)
+    tr_crit_vals     = np.zeros_like(Cp_vals)
     
-    tr_crit_top_func      = interp1d( x_t_top[::-1], tr_crit_top[::-1]) 
-    tr_crit_bot_func      = interp1d( x_t_bot, tr_crit_bot) 
-    Re_theta_t_top_func   = interp1d(x_t_top[::-1],  Re_theta_t_top[::-1])
-    Re_theta_t_bot_func   = interp1d(x_t_bot,Re_theta_t_bot)
-    Re_theta_t_vals       = np.concatenate([Re_theta_t_top_func(x_top[::-1]), Re_theta_t_bot_func(x_bot)])
-    tr_crit_vals          = np.concatenate([tr_crit_top_func(x_top[::-1]),tr_crit_bot_func(x_bot)])
      
-    s_vals                = np.zeros_like(x_vals)
-    H_star_vals           = np.zeros_like(x_vals)
-    P_vals                = np.zeros_like(x_vals)     
-    m_vals                = np.zeros_like(x_vals)     
-    K_vals                = np.zeros_like(x_vals)     
-    tau_vals              = np.zeros_like(x_vals)  
-    Di_val                = np.zeros_like(x_vals)   
-     
-    Cl,Cd,Cm = aero_coeff(x_vals,y_vals,Cp_vals,alpha,npanel)  
+    for ial in range(nalpha):
+        for iRe in range(nRe): 
+            # Find stagnation point             
+            i_stag  = np.where(vt[:,ial,iRe] > 0)[0][0]  
+            
+            # ---------------------------------------------------------------------
```

Comment:
I feel like you should be able to pull more of this code out of the loop

Reply:
I used the modulus operator to loop repeat loop

Reply:
I was able to use some interesting numpy functions include np.ma.maskwhere, np.mask and np.choose to vectorize parts of this double loop. However, there are some parts that require a for-loop. Unfortunately, unlike the acoustics branch, the transition location changes with AoA and Re so are of the code where I am concatenating create vectors of different arrays. This creates vectors of different lengths in one of the directions. After facing issues concatenation and interpolating over slices of a masked array I decided to keep the two loops. It is slow, but it is readable, which will be useful when I am doing acoustics in the future. 

Moreover, I made the code to do batch analysis but realized that sometimes, I just want to run a specific number of angle of attacks at a specific Re i.e, if I am analyzing a wing. As opposed to trying a diagonal masked array over **n^2** points, I simply reduce the dimension and only evaluate the points I need, essentially making the for loop of dimension **n**

Reply:
I did keep my attempt in the experimental branch...however it is not pretty at all